src/index.js

```diff
@@ -202,172 +286,266 @@ class PeerId {
+   * @returns {Promise<PeerId>}
+   */
+  static async createFromJSON (obj) {
+    const id = mh.fromB58String(obj.id)
+    const rawPrivKey = obj.privKey && uint8ArrayFromString(obj.privKey, 'base64pad')
+    const rawPubKey = obj.pubKey && uint8ArrayFromString(obj.pubKey, 'base64pad')
+    const pub = rawPubKey && await cryptoKeys.unmarshalPublicKey(rawPubKey)
+
+    if (!rawPrivKey) {
+      return new PeerId(id, undefined, pub)
+    }
 
-  const privKey = await cryptoKeys.unmarshalPrivateKey(key)
-  return computePeerId(privKey, privKey.public)
-}
+    const privKey = await cryptoKeys.unmarshalPrivateKey(rawPrivKey)
+    const privDigest = await computeDigest(privKey.public)
 
-exports.createFromJSON = async (obj) => {
-  const id = mh.fromB58String(obj.id)
-  const rawPrivKey = obj.privKey && uint8ArrayFromString(obj.privKey, 'base64pad')
-  const rawPubKey = obj.pubKey && uint8ArrayFromString(obj.pubKey, 'base64pad')
-  const pub = rawPubKey && await cryptoKeys.unmarshalPublicKey(rawPubKey)
+    /**
+     * @type {Uint8Array}
+     */
+    let pubDigest = new Uint8Array()
 
-  if (!rawPrivKey) {
-    return new PeerIdWithIs(id, null, pub)
-  }
+    if (pub) {
+      pubDigest = await computeDigest(pub)
+    }
 
-  const privKey = await cryptoKeys.unmarshalPrivateKey(rawPrivKey)
-  const privDigest = await computeDigest(privKey.public)
+    if (pub && !uint8ArrayEquals(privDigest, pubDigest)) {
+      throw new Error('Public and private key do not match')
+    }
 
-  let pubDigest
+    if (id && !uint8ArrayEquals(privDigest, id)) {
+      throw new Error('Id and private key do not match')
+    }
 
-  if (pub) {
-    pubDigest = await computeDigest(pub)
+    return new PeerId(id, privKey, pub)
   }
 
-  if (pub && !uint8ArrayEquals(privDigest, pubDigest)) {
-    throw new Error('Public and private key do not match')
-  }
+  /**
+   * Create PeerId from Protobuf bytes. 
+   * 
+   * @param {Uint8Array | string} buf - Protobuf bytes, as Uint8Array or hex-encoded string.
+   * @returns {Promise<PeerId>}
+   */
+  static async createFromProtobuf (buf) {
+    if (typeof buf === 'string') {
+      buf = uint8ArrayFromString(buf, 'base16')
+    }
 
-  if (id && !uint8ArrayEquals(privDigest, id)) {
-    throw new Error('Id and private key do not match')
-  }
+    let { id, privKey: privKeyArray, pubKey: pubKeyArray } = PeerIdProto.decode(buf)
 
-  return new PeerIdWithIs(id, privKey, pub)
-}
+    /**
+     * @type {PrivateKey | false}
+     */
+    const privKey = privKeyArray ? await cryptoKeys.unmarshalPrivateKey(privKeyArray) : false
 
-exports.createFromProtobuf = async (buf) => {
-  if (typeof buf === 'string') {
-    buf = uint8ArrayFromString(buf, 'base16')
-  }
+    /**
+     * @type {PublicKey | false}
+     */
+    const pubKey = pubKeyArray ? await cryptoKeys.unmarshalPublicKey(pubKeyArray) : false
 
-  let { id, privKey, pubKey } = PeerIdProto.decode(buf)
+    /**
+     * @type {Uint8Array}
+     */
+    let pubDigest = new Uint8Array()
```

Comment:
I think we should avoid this uint8array allocations. For ts to be happy, probably add that pubDigest exists in the if statements, if that is the problem